network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -161,7 +206,17 @@ impl<T: Ord> ConnectionState<T> {
     fn mark_banned(&mut self, id: T) {
         self.failed.remove(&id);
         self.down.remove(&id);
-        self.banned.insert(id);
+        self.banned.insert(id.clone());
+
+        let unban_deadline = Instant::now() + self.ban_time;
+        // If no peer is scheduled to be unbanned, put is as the first-to-be unbanned peer.
+        // Otherwise queue the peer such that it will be scheduled later when the peer is the first in line.
+        if self.unban_timeout.is_none() {
+            self.unban_timeout = Some((id, Box::pin(sleep_until(unban_deadline))));
+            self.waker.wake();
+        } else {
+            self.unban_deadlines.push_back((id.clone(), unban_deadline));
+        }
     }
 
     /// Removes a connection ID from the banned set
```

Comment:
The `unmark_banned` function below doesn't clean up the `self.unban_deadlines`. I'd suggest to just remove the function since it's currently unused.

Reply:
Removed `unmark_banned` and `unban_connection` including its unused caller and `NetworkAction`.